src/journal/sd-journal.c

```diff
@@ -2661,13 +2661,26 @@ _public_ int sd_journal_wait(sd_journal *j, uint64_t timeout_usec) {
         assert_return(!journal_pid_changed(j), -ECHILD);
 
         if (j->inotify_fd < 0) {
+                Iterator i;
+                JournalFile *f;
 
                 /* This is the first invocation, hence create the
                  * inotify watch */
                 r = sd_journal_get_fd(j);
                 if (r < 0)
                         return r;
 
+                /* Server might have done some vacuuming while we weren't watching.
+                   Get rid of the deleted files now so they don't stay around indefinitely. */
+                ORDERED_HASHMAP_FOREACH(f, j->files, i) {
+                        r = journal_file_fstat(f);
+                        if (r < 0)
+                                log_debug_errno(r,"Failed to fstat() journal file '%s' : %m", f->path);
```

Comment:
If the fstat failed, it seems wrong to continue and look that the stats.

Reply:
Well we did the `fstat()` when we've loaded the file initially. I think that looking at possibly outdated data (and having a minuscule risk of TOCTOU) is better than skipping the file altogether.

Reply:
We shouldn't do this. The Linux VFS layer is inherently racy, and it's entirely possible nlink can become >0 again later. Probably that won't happen a *lot* later unless we hit some pathological case where we blocked in kernel space for a while, but even that is possible.

If we can't stat it, we should err on the side of caution and do nothing.

Reply:
Ok, let's just skip the file then.